From: Kalpak Shah <kalpak@clusterfs.com>
This nanosecond patch seems to be missing the fix below which is required for http://bugzilla.kernel.org/show_bug.cgi?id=5079

If the timestamp is set to before epoch i.e. a negative timestamp then the file may have its date set into the future on 64-bit systems. So when the timestamp is read it must be cast as signed.

Signed-off-by: Kalpak Shah <kalpak@clusterfs.com>

---
 include/linux/ext4_fs.h |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

Index: linux-2.6.22/include/linux/ext4_fs.h
===================================================================
--- linux-2.6.22.orig/include/linux/ext4_fs.h	2007-07-12 13:20:02.000000000 -0700
+++ linux-2.6.22/include/linux/ext4_fs.h	2007-07-12 13:23:03.000000000 -0700
@@ -400,7 +400,7 @@ do {									       \
 
 #define EXT4_INODE_GET_XTIME(xtime, inode, raw_inode)			       \
 do {									       \
-	(inode)->xtime.tv_sec = le32_to_cpu((raw_inode)->xtime);	       \
+	(inode)->xtime.tv_sec = (signed)le32_to_cpu((raw_inode)->xtime);	       \
 	if (EXT4_FITS_IN_INODE(raw_inode, EXT4_I(inode), xtime ## _extra))     \
 		ext4_decode_extra_time(&(inode)->xtime,			       \
 				       raw_inode->xtime ## _extra);	       \
@@ -409,7 +409,7 @@ do {									       \
 #define EXT4_EINODE_GET_XTIME(xtime, einode, raw_inode)			       \
 do {									       \
 	if (EXT4_FITS_IN_INODE(raw_inode, einode, xtime))		       \
-		(einode)->xtime.tv_sec = le32_to_cpu((raw_inode)->xtime);      \
+		(einode)->xtime.tv_sec = (signed)le32_to_cpu((raw_inode)->xtime);      \
 	if (EXT4_FITS_IN_INODE(raw_inode, einode, xtime ## _extra))	       \
 		ext4_decode_extra_time(&(einode)->xtime,		       \
 				       raw_inode->xtime ## _extra);	       \
